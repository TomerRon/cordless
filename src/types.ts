import { APIApplicationCommandOptionChoice } from 'discord-api-types/v10'
import Discord, {
  ActionRowBuilder,
  ApplicationCommandOptionAllowedChannelTypes,
  ApplicationCommandOptionType,
  ButtonBuilder,
  ButtonInteraction,
  ButtonStyle,
  ChatInputCommandInteraction,
  ClientEvents,
  ClientOptions,
  InteractionResponse,
} from 'discord.js'

/** Initialization options for your cordless bot */
export type InitOptions<C extends CustomContext = {}> = {
  /**
   * Your bot token.
   *
   * @see https://discordjs.guide/preparations/setting-up-a-bot-application.html#your-bot-s-token
   */
  token: string
  /** The commands used by your bot. */
  commands?: BotCommand<C>[]
  /** The event handlers used by your bot. */
  handlers?: BotEventHandler<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
  /** A custom context object which will extend the context passed to your commands and event handlers */
  context?: C
  /**
   * Override the default Gateway Intents of the discord.js client.
   *
   * By default, the discord.js client will initialize with the [GUILDS, GUILD_MESSAGES] intents.
   * These default intents should be sufficient in most cases.
   *
   * @see https://discord.com/developers/docs/topics/gateway#gateway-intents
   * @see https://discordjs.guide/popular-topics/intents.html
   */
  intents?: ClientOptions['intents']
}

export type BotCommand<C extends CustomContext = {}> =
  | BotCommandWithHandler<C>
  | BotCommandWithSubcommands<C>

export interface BotCommandWithHandler<C extends CustomContext = {}>
  extends BotCommandBase {
  handler: (
    args: BotCommandHandlerArgs<C>,
  ) => void | Promise<InteractionResponse<boolean> | void>
  components?: BotCommandComponent<C>[]
  options?: BotCommandOption[]
  subcommands?: never
}

export type BotCommandHandlerArgs<C extends CustomContext = {}> = {
  interaction: ChatInputCommandInteraction
  context: Context<C>
  components?: ActionRowBuilder<ButtonBuilder>[]
}

export type BotCommandComponent<C extends CustomContext = {}> =
  | BotCommandButtonComponent<C>
  | BotCommandLinkComponent<C>

export interface BotCommandButtonComponent<C extends CustomContext = {}>
  extends BotCommandButtonComponentBase {
  style?: Exclude<ButtonStyle, ButtonStyle.Link>
  handler: BotCommandButtonHandler<C>
}

export type BotCommandButtonHandler<C extends CustomContext = {}> = (
  args: BotCommandButtonHandlerArgs<C>,
) => void | Promise<InteractionResponse<boolean> | void>

export interface BotCommandLinkComponent<C extends CustomContext = {}>
  extends BotCommandButtonComponentBase {
  style: ButtonStyle.Link
  url:
    | string
    | ((
        args: Omit<BotCommandHandlerArgs<C>, 'components'>,
      ) => string | Promise<string>)
}

type BotCommandButtonComponentBase = {
  label: string
  style?: ButtonStyle
}

export type BotCommandButtonHandlerArgs<C extends CustomContext = {}> = {
  interaction: ButtonInteraction
  context: Context<C>
}

export interface BotCommandWithSubcommands<C extends CustomContext = {}>
  extends BotCommandBase {
  subcommands: BotCommandWithHandler<C>[]
  handler?: never
  options?: never
}

type BotCommandBase = {
  name: string
  description?: string
}

export type BotCommandOption =
  | BotCommandStringOption
  | BotCommandIntegerOption
  | BotCommandBooleanOption
  | BotCommandUserOption
  | BotCommandChannelOption
  | BotCommandRoleOption
  | BotCommandMentionableOption
  | BotCommandNumberOption
  | BotCommandAttachmentOption

export interface BotCommandStringOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.String
  choices?: APIApplicationCommandOptionChoice<string>[]
}

export interface BotCommandIntegerOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Integer
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandBooleanOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Boolean
}

export interface BotCommandUserOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.User
}

export interface BotCommandChannelOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Channel
  channelTypes?: ApplicationCommandOptionAllowedChannelTypes[]
}

export interface BotCommandRoleOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Role
}

export interface BotCommandMentionableOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Mentionable
}

export interface BotCommandNumberOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Number
  choices?: APIApplicationCommandOptionChoice<number>[]
  min?: number
  max?: number
}

export interface BotCommandAttachmentOption extends BotCommandOptionBase {
  type: ApplicationCommandOptionType.Attachment
}

type BotCommandOptionBase = {
  type: ApplicationCommandOptionType
  name: string
  description?: string
  required?: boolean
}

export type BotEventHandler<
  E extends keyof ClientEvents = 'messageCreate',
  C extends CustomContext = {},
> = {
  /**
   * The event this handler should subscribe to (default: "messageCreate").
   */
  event?: E
  /** Determines whether or not the callback should run */
  condition: (...args: [...ClientEvents[E], Context<C>]) => boolean
  /** Called whenever an event that matches the condition is received */
  callback: (
    ...args: [...ClientEvents[E], Context<C>]
  ) => void | Promise<Discord.Message | void>
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type CustomContext = Record<string, any>

export type Context<C extends CustomContext = {}> = {
  client: Discord.Client<true>
  handlers: BotEventHandler<any, C>[] // eslint-disable-line @typescript-eslint/no-explicit-any
} & C
