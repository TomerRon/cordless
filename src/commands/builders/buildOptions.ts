import {
  ApplicationCommandOptionType,
  SlashCommandBuilder,
  SlashCommandSubcommandBuilder,
} from 'discord.js'
import { BotCommandOption } from '../../types'

/**
 * Adds a list of options to a SlashCommand
 */
const buildOptions = (
  cmd: SlashCommandBuilder | SlashCommandSubcommandBuilder,
  options: BotCommandOption[],
) => {
  options.forEach((option) => {
    const {
      type,
      name,
      description = 'No description',
      required = false,
    } = option

    switch (type) {
      case ApplicationCommandOptionType.String:
        cmd.addStringOption((o) => {
          o.setName(name).setDescription(description).setRequired(required)

          if (option.choices) {
            o.addChoices(...option.choices)
          }

          return o
        })
        break

      case ApplicationCommandOptionType.Integer:
        cmd.addIntegerOption((o) => {
          o.setName(name).setDescription(description).setRequired(required)

          if (option.choices) {
            o.addChoices(...option.choices)
          }

          if (option.min) {
            o.setMinValue(option.min)
          }

          if (option.max) {
            o.setMaxValue(option.max)
          }

          return o
        })
        break

      case ApplicationCommandOptionType.Boolean:
        cmd.addBooleanOption((o) =>
          o.setName(name).setDescription(description).setRequired(required),
        )
        break

      case ApplicationCommandOptionType.User:
        cmd.addUserOption((o) =>
          o.setName(name).setDescription(description).setRequired(required),
        )
        break

      case ApplicationCommandOptionType.Channel:
        cmd.addChannelOption((o) => {
          o.setName(name).setDescription(description).setRequired(required)

          if (option.channelTypes) {
            o.addChannelTypes(...option.channelTypes)
          }

          return o
        })
        break

      case ApplicationCommandOptionType.Role:
        cmd.addRoleOption((o) =>
          o.setName(name).setDescription(description).setRequired(required),
        )
        break

      case ApplicationCommandOptionType.Mentionable:
        cmd.addMentionableOption((o) =>
          o.setName(name).setDescription(description).setRequired(required),
        )
        break

      case ApplicationCommandOptionType.Number:
        cmd.addNumberOption((o) => {
          o.setName(name).setDescription(description).setRequired(required)

          if (option.choices) {
            o.addChoices(...option.choices)
          }

          if (option.min) {
            o.setMinValue(option.min)
          }

          if (option.max) {
            o.setMaxValue(option.max)
          }

          return o
        })
        break

      case ApplicationCommandOptionType.Attachment:
        cmd.addAttachmentOption((o) =>
          o.setName(name).setDescription(description).setRequired(required),
        )
        break
    }
  })
}

export default buildOptions
