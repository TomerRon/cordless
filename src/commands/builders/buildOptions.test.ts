import {
  SlashCommandBuilder,
  SlashCommandRoleOption,
} from '@discordjs/builders'
import { ChannelType } from 'discord-api-types/v10'
import { BotCommandOption } from '../../types'
import buildOptions from './buildOptions'

describe('buildOptions', () => {
  beforeEach(jest.clearAllMocks)

  const mockCmd: Partial<Record<keyof SlashCommandBuilder, jest.Mock>> = {
    addStringOption: jest.fn().mockReturnThis(),
    addIntegerOption: jest.fn().mockReturnThis(),
    addBooleanOption: jest.fn().mockReturnThis(),
    addUserOption: jest.fn().mockReturnThis(),
    addChannelOption: jest.fn().mockReturnThis(),
    addRoleOption: jest.fn().mockReturnThis(),
    addMentionableOption: jest.fn().mockReturnThis(),
    addNumberOption: jest.fn().mockReturnThis(),
    addAttachmentOption: jest.fn().mockReturnThis(),
  }

  const mockBuilder: Record<string, jest.Mock> = {
    setName: jest.fn().mockReturnThis(),
    setDescription: jest.fn().mockReturnThis(),
    setRequired: jest.fn().mockReturnThis(),
    addChoices: jest.fn().mockReturnThis(),
    addChannelTypes: jest.fn().mockReturnThis(),
    setMinValue: jest.fn().mockReturnThis(),
    setMaxValue: jest.fn().mockReturnThis(),
  }

  describe('when there are no options', () => {
    it('does not add any options to the command', () => {
      buildOptions(mockCmd as unknown as SlashCommandBuilder, [])

      Object.values(mockCmd).forEach((mock) =>
        expect(mock).not.toHaveBeenCalled(),
      )
    })
  })

  describe('when one option is passed', () => {
    const option: BotCommandOption = {
      type: 'ROLE',
      name: 'mock-option',
      description: 'mock-description',
    }

    it('adds the relevant option to the command', () => {
      buildOptions(mockCmd as unknown as SlashCommandBuilder, [option])

      const { addRoleOption, ...rest } = mockCmd

      Object.values(rest).forEach((mock) => expect(mock).not.toHaveBeenCalled())

      expect(addRoleOption).toHaveBeenCalledWith(expect.any(Function))

      const callback = addRoleOption?.mock.calls[0][0] as (
        o: SlashCommandRoleOption,
      ) => SlashCommandRoleOption

      callback(mockBuilder as unknown as SlashCommandRoleOption)

      expect(mockBuilder.setName).toHaveBeenCalledWith(option.name)
      expect(mockBuilder.setDescription).toHaveBeenCalledWith(
        option.description,
      )
      expect(mockBuilder.setRequired).toHaveBeenCalledWith(false)
    })
  })

  describe('when multiple options are passed', () => {
    const optionA: BotCommandOption = {
      type: 'STRING',
      name: 'mock-option-a',
      description: 'mock-description-a',
      required: true,
      choices: [
        {
          name: 'choice-1-name',
          value: 'choice-1-value',
        },
        {
          name: 'choice-2-name',
          value: 'choice-2-value',
        },
      ],
    }

    const optionB: BotCommandOption = {
      type: 'STRING',
      name: 'mock-option-b',
    }

    const optionC: BotCommandOption = {
      type: 'NUMBER',
      name: 'mock-option-c',
      min: 5,
      max: 50,
    }

    const options = [optionA, optionB, optionC]

    beforeEach(() => {
      buildOptions(mockCmd as unknown as SlashCommandBuilder, options)
    })

    it('adds option A to the command', () => {
      expect(mockCmd.addStringOption).toHaveBeenNthCalledWith(
        1,
        expect.any(Function),
      )

      const callback = mockCmd.addStringOption?.mock.calls[0][0] as jest.Mock

      callback(mockBuilder)

      expect(mockBuilder.setName).toHaveBeenNthCalledWith(1, optionA.name)
      expect(mockBuilder.setDescription).toHaveBeenNthCalledWith(
        1,
        optionA.description,
      )
      expect(mockBuilder.setRequired).toHaveBeenNthCalledWith(
        1,
        optionA.required,
      )
      expect(mockBuilder.addChoices).toHaveBeenNthCalledWith(
        1,
        ...optionA.choices,
      )
    })

    it('adds option B to the command', () => {
      expect(mockCmd.addStringOption).toHaveBeenNthCalledWith(
        2,
        expect.any(Function),
      )

      const callback = mockCmd.addStringOption?.mock.calls[1][0] as jest.Mock

      callback(mockBuilder)

      expect(mockBuilder.setName).toHaveBeenNthCalledWith(1, optionB.name)
      expect(mockBuilder.setDescription).toHaveBeenNthCalledWith(
        1,
        'No description',
      )
      expect(mockBuilder.setRequired).toHaveBeenNthCalledWith(1, false)
      expect(mockBuilder.addChoices).not.toHaveBeenCalled()
    })

    it('adds option C to the command', () => {
      expect(mockCmd.addNumberOption).toHaveBeenNthCalledWith(
        1,
        expect.any(Function),
      )

      const callback = mockCmd.addNumberOption?.mock.calls[0][0] as jest.Mock

      callback(mockBuilder)

      expect(mockBuilder.setName).toHaveBeenNthCalledWith(1, optionC.name)
      expect(mockBuilder.setDescription).toHaveBeenNthCalledWith(
        1,
        'No description',
      )
      expect(mockBuilder.setRequired).toHaveBeenNthCalledWith(1, false)
      expect(mockBuilder.setMinValue).toHaveBeenNthCalledWith(1, optionC.min)
      expect(mockBuilder.setMaxValue).toHaveBeenNthCalledWith(1, optionC.max)
    })

    it('does not add any other options', () => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { addStringOption, addNumberOption, ...rest } = mockCmd

      Object.values(rest).forEach((mock) => expect(mock).not.toHaveBeenCalled())
    })
  })

  describe('snapshots', () => {
    const options: BotCommandOption[] = [
      {
        type: 'STRING',
        name: 'mock-option-a',
      },
      {
        type: 'STRING',
        name: 'mock-option-b',
        description: 'mock-descrption-b',
        required: true,
        choices: [
          {
            name: 'choice-1-name',
            value: 'choice-1-value',
          },
          {
            name: 'choice-2-name',
            value: 'choice-2-value',
          },
        ],
      },
      {
        type: 'INTEGER',
        name: 'mock-option-c',
      },
      {
        type: 'INTEGER',
        name: 'mock-option-d',
        description: 'mock-descrption-d',
        required: true,
        min: 10,
        max: 1337,
        choices: [
          {
            name: 'choice-1-name',
            value: 12,
          },
          {
            name: 'choice-2-name',
            value: 345,
          },
        ],
      },
      {
        type: 'BOOLEAN',
        name: 'mock-option-e',
      },
      {
        type: 'BOOLEAN',
        name: 'mock-option-f',
        description: 'mock-descrption-f',
        required: true,
      },
      {
        type: 'USER',
        name: 'mock-option-g',
      },
      {
        type: 'USER',
        name: 'mock-option-h',
        description: 'mock-descrption-h',
        required: true,
      },
      {
        type: 'CHANNEL',
        name: 'mock-option-i',
      },
      {
        type: 'CHANNEL',
        name: 'mock-option-j',
        description: 'mock-descrption-j',
        required: true,
        channelTypes: [ChannelType.GuildText, ChannelType.GuildVoice],
      },
      {
        type: 'ROLE',
        name: 'mock-option-k',
      },
      {
        type: 'ROLE',
        name: 'mock-option-l',
        description: 'mock-descrption-l',
        required: true,
      },
      {
        type: 'MENTIONABLE',
        name: 'mock-option-m',
      },
      {
        type: 'MENTIONABLE',
        name: 'mock-option-n',
        description: 'mock-descrption-n',
        required: true,
      },
      {
        type: 'NUMBER',
        name: 'mock-option-o',
      },
      {
        type: 'NUMBER',
        name: 'mock-option-p',
        description: 'mock-descrption-p',
        required: true,
        min: 10,
        max: 1337,
        choices: [
          {
            name: 'choice-1-name',
            value: 12,
          },
          {
            name: 'choice-2-name',
            value: 345,
          },
        ],
      },
      {
        type: 'ATTACHMENT',
        name: 'mock-option-q',
      },
      {
        type: 'ATTACHMENT',
        name: 'mock-option-r',
        description: 'mock-descrption-r',
        required: true,
      },
    ]

    it('matches the snapshot', () => {
      buildOptions(mockCmd as unknown as SlashCommandBuilder, options)

      Object.values(mockCmd).forEach((fn) => {
        expect(fn).toHaveBeenNthCalledWith(1, expect.any(Function))
        expect(fn).toHaveBeenNthCalledWith(2, expect.any(Function))

        const callbackA = fn.mock.calls[0][0] as jest.Mock

        callbackA(mockBuilder)

        const callbackB = fn.mock.calls[1][0] as jest.Mock

        callbackB(mockBuilder)
      })

      Object.entries(mockBuilder).forEach(([key, fn]) =>
        expect(fn.mock.calls).toMatchSnapshot(key),
      )
    })
  })
})
