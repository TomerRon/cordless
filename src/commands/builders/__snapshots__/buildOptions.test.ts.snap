// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildOptions snapshots matches the snapshot: addChannelTypes 1`] = `
Array [
  Array [
    0,
    2,
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: addChoices 1`] = `
Array [
  Array [
    Object {
      "name": "choice-1-name",
      "value": "choice-1-value",
    },
    Object {
      "name": "choice-2-name",
      "value": "choice-2-value",
    },
  ],
  Array [
    Object {
      "name": "choice-1-name",
      "value": 12,
    },
    Object {
      "name": "choice-2-name",
      "value": 345,
    },
  ],
  Array [
    Object {
      "name": "choice-1-name",
      "value": 12,
    },
    Object {
      "name": "choice-2-name",
      "value": 345,
    },
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: setDescription 1`] = `
Array [
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-b",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-d",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-f",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-h",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-j",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-l",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-n",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-p",
  ],
  Array [
    "No description",
  ],
  Array [
    "mock-descrption-r",
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: setMaxValue 1`] = `
Array [
  Array [
    1337,
  ],
  Array [
    1337,
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: setMinValue 1`] = `
Array [
  Array [
    10,
  ],
  Array [
    10,
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: setName 1`] = `
Array [
  Array [
    "mock-option-a",
  ],
  Array [
    "mock-option-b",
  ],
  Array [
    "mock-option-c",
  ],
  Array [
    "mock-option-d",
  ],
  Array [
    "mock-option-e",
  ],
  Array [
    "mock-option-f",
  ],
  Array [
    "mock-option-g",
  ],
  Array [
    "mock-option-h",
  ],
  Array [
    "mock-option-i",
  ],
  Array [
    "mock-option-j",
  ],
  Array [
    "mock-option-k",
  ],
  Array [
    "mock-option-l",
  ],
  Array [
    "mock-option-m",
  ],
  Array [
    "mock-option-n",
  ],
  Array [
    "mock-option-o",
  ],
  Array [
    "mock-option-p",
  ],
  Array [
    "mock-option-q",
  ],
  Array [
    "mock-option-r",
  ],
]
`;

exports[`buildOptions snapshots matches the snapshot: setRequired 1`] = `
Array [
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
]
`;
